# <b><font color=dodgerblue> Python Worksheet 3
# <b><font color=dodgerblue> Melbourne-House-Prices

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
import copy

In [2]:
Original_data = pd.read_csv("Melbourne-House-Prices.csv")

df = copy.deepcopy(Original_data)
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[23678 rows x 21 columns]

# <b><font color=dodgerblue> EDA

### <b><font color=dodgerblue> Q1. Determine the no. of rows and no. of columns in the dataset
### Ans :-

In [3]:
df.shape

(23678, 21)

### <b><font color=dodgerblue> Q2. Display the datatypes of each column
### Ans :-

In [4]:
df.dtypes.to_frame()

0
Suburb          object
Address         object
Rooms          float64
Type            object
Price          float64
Method          object
SellerG         object
Date            object
Distance       float64
Postcode       float64
Bedroom2       float64
Bathroom       float64
Car            float64
Landsize       float64
BuildingArea   float64
YearBuilt      float64
CouncilArea     object
Lattitude      float64
Longtitude     float64
Regionname      object
Propertycount  float64

### OR

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23678 entries, 0 to 23677
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  float64
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

#### From counting of values, we can see the dataset have any missing values or not

### <b><font color=dodgerblue> Q3. Determine the row indexes of rows having all null values (i.e. empty rows)
### Ans :-

In [6]:
df.isnull().sum(axis=0).sort_values()

Suburb             131
Address            131
Rooms              131
Type               131
Method             131
SellerG            131
Date               131
Regionname         132
Postcode           132
Propertycount      132
Distance           132
Lattitude         4435
Longtitude        4435
Bedroom2          4612
Bathroom          4615
Car               4757
Price             5282
Landsize          6268
CouncilArea       8022
YearBuilt        12138
BuildingArea     13660
dtype: int64

### <b><font color=dodgerblue> Q4. Determine the total number of empty rows and their percentage in whole dataset and remove them.


In [7]:
df.isnull().sum(axis=1).sort_values(ascending=False)

5328     21
4897     21
16252    21
17246    21
18077    21
         ..
5078      0
19077     0
19078     0
5081      0
8555      0
Length: 23678, dtype: int64

### Rows having all misssing values


In [8]:
df.isnull().all(axis=1).sum()

131

#### We have 131 total rows, with all missing value in a row, So we will remove those row

In [9]:
null_columns=df.columns[df.isnull().any()]
df=df[null_columns].dropna(how='all')
print("New Data Shape",df.shape)

New Data Shape (23547, 21)


In [10]:
df.shape

(23547, 21)

In [11]:
print("Percentage of all NaN values to the whole data set ==>",(131/23678) * 100)

Percentage of all NaN values to the whole data set ==> 0.5532561871779711


### <b><font color=dodgerblue> Q5. Determine the column names having missing values.
### Ans :-

In [12]:
df.isnull().sum() > 0

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [13]:
df.isnull().any()

Suburb           False
Address          False
Rooms            False
Type             False
Price             True
Method           False
SellerG          False
Date             False
Distance          True
Postcode          True
Bedroom2          True
Bathroom          True
Car               True
Landsize          True
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude         True
Longtitude        True
Regionname        True
Propertycount     True
dtype: bool

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23547 entries, 0 to 23677
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  float64
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

### <b><font color=dodgerblue> Q6. Determine the missing values and their percentage in each column.
### Ans :-

In [15]:
round(df.isnull().sum().sort_values(ascending=False)/len(df)*100,2)

BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Landsize         26.06
Price            21.88
Car              19.65
Bathroom         19.04
Bedroom2         19.03
Longtitude       18.28
Lattitude        18.28
Distance          0.00
Propertycount     0.00
Postcode          0.00
Regionname        0.00
Date              0.00
SellerG           0.00
Method            0.00
Type              0.00
Rooms             0.00
Address           0.00
Suburb            0.00
dtype: float64

### OR

In [16]:
col_names =df.columns
num_df = df.shape[0]

for c in col_names:
        num_non = df[c].isin(["NaN"]).sum()
        if num_non > 0:
            print(c)
            print(num_non)
            print("{0:.2f}%".format(float(num_non)/ num_df * 100))

Price
5151
21.88%
Distance
1
0.00%
Postcode
1
0.00%
Bedroom2
4481
19.03%
Bathroom
4484
19.04%
Car
4626
19.65%
Landsize
6137
26.06%
BuildingArea
13529
57.46%
YearBuilt
12007
50.99%
Lattitude
4304
18.28%
Longtitude
4304
18.28%
Propertycount
1
0.00%


### <b><font color=dodgerblue> Q7. Determine the row indexes having missing values.
### Ans :-

In [17]:
df.isnull().any(axis=1)

0         True
1         True
2        False
3         True
4        False
         ...  
23673     True
23674     True
23675     True
23676     True
23677     True
Length: 23547, dtype: bool

### Rows having all misssing values

In [18]:
df.isnull().all(axis=1).sum()

0

### <b><font color=dodgerblue> Q8. Determine the missing values and their percentage in each row.
### Ans :-

In [19]:
row_missing = df.isnull().sum(axis=1) * 100 / 21
row_missing.sort_values(ascending=False)

18630    61.904762
17198    47.619048
16326    47.619048
21089    47.619048
6348     47.619048
           ...    
11629     0.000000
5097      0.000000
11627     0.000000
11626     0.000000
8606      0.000000
Length: 23547, dtype: float64

### <b><font color=dodgerblue> Q9. Determine the column names having more than 30% missing values and remove them from the dataset.
### Ans :-

In [20]:
col_per = df.isnull().sum().sort_values(ascending=False)/len(df)*100
col_per

BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Landsize         26.062768
Price            21.875398
Car              19.645815
Bathroom         19.042766
Bedroom2         19.030025
Longtitude       18.278337
Lattitude        18.278337
Distance          0.004247
Propertycount     0.004247
Postcode          0.004247
Regionname        0.004247
Date              0.000000
SellerG           0.000000
Method            0.000000
Type              0.000000
Rooms             0.000000
Address           0.000000
Suburb            0.000000
dtype: float64

In [21]:
col = df.isnull().sum().sort_values(ascending=False).head(3).index.values
col

array(['BuildingArea', 'YearBuilt', 'CouncilArea'], dtype=object)

In [22]:
df.drop(col,axis="columns",inplace=True)
print("New Data Shape",df.shape)

New Data Shape (23547, 18)


In [23]:
print("Missing values Percentage wise in all new data","\n",df.isnull().sum().sort_values(ascending=False)/len(df)*100)

Missing values Percentage wise in all new data 
 Landsize         26.062768
Price            21.875398
Car              19.645815
Bathroom         19.042766
Bedroom2         19.030025
Longtitude       18.278337
Lattitude        18.278337
Propertycount     0.004247
Regionname        0.004247
Distance          0.004247
Postcode          0.004247
Date              0.000000
SellerG           0.000000
Method            0.000000
Type              0.000000
Rooms             0.000000
Address           0.000000
Suburb            0.000000
dtype: float64


### <b><font color=dodgerblue> Q10. Determine the number and percentage of rows having more than 5 missing values. If the percentage of suchrows is less than 20% then remove all these rows.
### Ans :-

In [24]:
df[df.isnull().sum(axis=1) > 5]

Suburb              Address  Rooms Type      Price Method  \
15            Abbotsford     217 Langridge St    3.0    h  1000000.0      S   
16            Abbotsford      18a Mollison St    2.0    t   745000.0      S   
19            Abbotsford  403/609 Victoria St    2.0    u   542000.0      S   
21            Abbotsford    25/84 Trenerry Cr    2.0    u   760000.0     SP   
22            Abbotsford    106/119 Turner St    1.0    u   481000.0     SP   
...                  ...                  ...    ...  ...        ...    ...   
23626  Templestowe Lower  1/207 Manningham Rd    2.0    u   550000.0     PI   
23630          Thornbury      1/128 Dundas St    2.0    t   770000.0     PI   
23633          Thornbury        111 Pender St    2.0    t   858000.0      S   
23639             Toorak   21/1059 Malvern Rd    2.0    u   720000.0     VB   
23662           Werribee         91 Latham St    4.0    h   540000.0     PI   

            SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
15           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
16           Jellis  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
19           Dingle  08-10-2016       2.5    3067.0       NaN       NaN  NaN   
21           Biggin  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
22     Purplebricks  10-12-2016       2.5    3067.0       NaN       NaN  NaN   
...             ...         ...       ...       ...       ...       ...  ...   
23626         Barry  26-08-2017      12.4    3107.0       NaN       NaN  NaN   
23630       McGrath  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23633        Jellis  26-08-2017       7.0    3071.0       NaN       NaN  NaN   
23639        Beller  26-08-2017       4.1    3142.0       NaN       NaN  NaN   
23662       Triwest  26-08-2017      14.7    3030.0       NaN       NaN  NaN   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
15          NaN        NaN         NaN  Northern Metropolitan         4019.0  
16          NaN        NaN         NaN  Northern Metropolitan         4019.0  
19          NaN        NaN         NaN  Northern Metropolitan         4019.0  
21          NaN        NaN         NaN  Northern Metropolitan         4019.0  
22          NaN        NaN         NaN  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23626       NaN        NaN         NaN   Eastern Metropolitan         5420.0  
23630       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23633       NaN        NaN         NaN  Northern Metropolitan         8870.0  
23639       NaN        NaN         NaN  Southern Metropolitan         7217.0  
23662       NaN        NaN         NaN   Western Metropolitan        16166.0  

[4278 rows x 18 columns]

In [25]:
print("Percentage of Number of rows having more than 5 missing values :-",len(df[df.isnull().sum(axis=1) > 5])/len(df)*100)

Percentage of Number of rows having more than 5 missing values :- 18.16791948018856


##### So we are going to remove these rows

In [26]:
df.isnull().sum(axis=1) <=5

0        True
1        True
2        True
3        True
4        True
         ... 
23673    True
23674    True
23675    True
23676    True
23677    True
Length: 23547, dtype: bool

In [27]:
df[df.isnull().sum(axis=1) <=5]

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23674       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19269 rows x 18 columns]

##### just seeing the data, if we remove the rows

In [28]:
df = df[df.isnull().sum(axis=1) <= 5]
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23673     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23674       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[19269 rows x 18 columns]

##### Now we have removed that rows, which having more than 5 missing values.

### <b><font color=dodgerblue> Q11. Display the description of data and percentage of missing values still left column wise.
### Ans :-

In [29]:
df.describe().T

count          mean            std          min          25%  \
Rooms          19269.0  2.973117e+00       0.978969      1.00000       2.0000   
Price          15086.0  1.066695e+06  638933.149010  85000.00000  640000.0000   
Distance       19269.0  1.020436e+01       6.065786      0.00000       6.1000   
Postcode       19269.0  3.110016e+03      93.730487   3000.00000    3049.0000   
Bedroom2       19066.0  2.951956e+00       0.996032      0.00000       2.0000   
Bathroom       19063.0  1.570897e+00       0.712684      0.00000       1.0000   
Car            18921.0  1.626235e+00       0.974048      0.00000       1.0000   
Landsize       17410.0  5.517835e+02    3544.288014      0.00000     181.0000   
Lattitude      19243.0 -3.781243e+01       0.079926    -38.18418     -37.8593   
Longtitude     19243.0  1.450003e+02       0.106071    144.43162     144.9393   
Propertycount  19269.0  7.490250e+03    4319.566673    129.00000    4442.0000   

                       50%           75%           max  
Rooms               3.0000  4.000000e+00  1.200000e+01  
Price          900000.0000  1.320000e+06  9.000000e+06  
Distance            9.2000  1.300000e+01  4.810000e+01  
Postcode         3102.0000  3.150000e+03  3.977000e+03  
Bedroom2            3.0000  4.000000e+00  3.000000e+01  
Bathroom            1.0000  2.000000e+00  1.200000e+01  
Car                 2.0000  2.000000e+00  2.600000e+01  
Landsize          448.0000  6.560000e+02  4.330140e+05  
Lattitude         -37.8097 -3.775980e+01 -3.740758e+01  
Longtitude        145.0043  1.450631e+02  1.455264e+02  
Propertycount    6567.0000  1.033100e+04  2.165000e+04

In [30]:
df.isnull().sum().sort_values(ascending=False)/len(df) *100

Price            21.708444
Landsize          9.647621
Car               1.806010
Bathroom          1.069075
Bedroom2          1.053506
Longtitude        0.134932
Lattitude         0.134932
Propertycount     0.000000
Method            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Distance          0.000000
SellerG           0.000000
Date              0.000000
Regionname        0.000000
Postcode          0.000000
Suburb            0.000000
dtype: float64

### <b><font color=dodgerblue> Q12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? Give a brief explanation and treat the missing values in 'Price' column.
### Ans :-

# Method 1
### We will remove all the rows, where Price values is missing

In [31]:
df1 = copy.deepcopy(df)
df1.shape

(19269, 18)

In [32]:
df1[df1["Price"].notnull()]

Suburb           Address  Rooms Type      Price Method  \
1        Abbotsford      85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford   25 Bloomburg St    2.0    h  1035000.0      S   
4        Abbotsford      5 Charles St    3.0    h  1465000.0     SP   
5        Abbotsford  40 Federation La    3.0    h   850000.0     PI   
6        Abbotsford       55a Park St    4.0    h  1600000.0     VB   
...             ...               ...    ...  ...        ...    ...   
23671  Williamstown   8/2 Thompson St    2.0    t   622500.0     SP   
23672  Williamstown      96 Verdon St    4.0    h  2500000.0     PI   
23675     Yallambie      17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville        6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville     33 Freeman St    4.0    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6          Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...           ...         ...       ...       ...       ...       ...  ...   
23671        Greg  26-08-2017       6.8    3016.0       2.0       2.0  1.0   
23672     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23675  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
5          94.0  -37.79690   144.99690  Northern Metropolitan         4019.0  
6         120.0  -37.80720   144.99410  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23671       NaN  -37.86393   144.90484   Western Metropolitan         6380.0  
23672     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[15086 rows x 18 columns]

##### Just seeing data, if we remove the missing values from Price attributes

In [33]:
df1 = df1[df1["Price"].notnull()]
df1

Suburb           Address  Rooms Type      Price Method  \
1        Abbotsford      85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford   25 Bloomburg St    2.0    h  1035000.0      S   
4        Abbotsford      5 Charles St    3.0    h  1465000.0     SP   
5        Abbotsford  40 Federation La    3.0    h   850000.0     PI   
6        Abbotsford       55a Park St    4.0    h  1600000.0     VB   
...             ...               ...    ...  ...        ...    ...   
23671  Williamstown   8/2 Thompson St    2.0    t   622500.0     SP   
23672  Williamstown      96 Verdon St    4.0    h  2500000.0     PI   
23675     Yallambie      17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville        6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville     33 Freeman St    4.0    h  1050000.0     VB   

          SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1          Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2          Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5          Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6          Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...           ...         ...       ...       ...       ...       ...  ...   
23671        Greg  26-08-2017       6.8    3016.0       2.0       2.0  1.0   
23672     Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23675  Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676     Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677     Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
5          94.0  -37.79690   144.99690  Northern Metropolitan         4019.0  
6         120.0  -37.80720   144.99410  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23671       NaN  -37.86393   144.90484   Western Metropolitan         6380.0  
23672     866.0  -37.85908   144.89299   Western Metropolitan         6380.0  
23675       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23677       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[15086 rows x 18 columns]

##### Now we have removed that rows , which are having missing Prices values

### <b><font color=dodgerblue> Q13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most suitable method.
### Ans :-

In [34]:
df1.isnull().sum().sort_values(ascending=False)/len(df1)*100

Landsize         9.830306
Car              1.763224
Bathroom         1.067215
Bedroom2         1.053957
Longtitude       0.145831
Lattitude        0.145831
Propertycount    0.000000
Method           0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Distance         0.000000
SellerG          0.000000
Date             0.000000
Regionname       0.000000
Postcode         0.000000
Suburb           0.000000
dtype: float64

In [35]:
df1["Landsize"].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

In [36]:
df1["Landsize"].notnull()

1         True
2         True
4         True
5         True
6         True
         ...  
23671    False
23672     True
23675    False
23676     True
23677    False
Name: Landsize, Length: 15086, dtype: bool

In [37]:
df1 = df1[df1["Landsize"].notnull()]
df1

Suburb           Address  Rooms Type      Price Method  \
1         Abbotsford      85 Turner St    2.0    h  1480000.0      S   
2         Abbotsford   25 Bloomburg St    2.0    h  1035000.0      S   
4         Abbotsford      5 Charles St    3.0    h  1465000.0     SP   
5         Abbotsford  40 Federation La    3.0    h   850000.0     PI   
6         Abbotsford       55a Park St    4.0    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
23668  Wheelers Hill      12 Strada Cr    4.0    h  1245000.0      S   
23669   Williamstown     77 Merrett Dr    3.0    h  1031000.0     SP   
23670   Williamstown       83 Power St    3.0    h  1170000.0      S   
23672   Williamstown      96 Verdon St    4.0    h  2500000.0     PI   
23676     Yarraville        6 Agnes St    4.0    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1        Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2        Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4        Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5        Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6        Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...         ...         ...       ...       ...       ...       ...  ...   
23668     Barry  26-08-2017      16.7    3150.0       4.0       2.0  2.0   
23669  Williams  26-08-2017       6.8    3016.0       3.0       2.0  2.0   
23670     Raine  26-08-2017       6.8    3016.0       3.0       2.0  4.0   
23672   Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23676   Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   

       Landsize  Lattitude  Longtitude                  Regionname  \
1         202.0  -37.79960   144.99840       Northern Metropolitan   
2         156.0  -37.80790   144.99340       Northern Metropolitan   
4         134.0  -37.80930   144.99440       Northern Metropolitan   
5          94.0  -37.79690   144.99690       Northern Metropolitan   
6         120.0  -37.80720   144.99410       Northern Metropolitan   
...         ...        ...         ...                         ...   
23668     652.0  -37.90562   145.16761  South-Eastern Metropolitan   
23669     333.0  -37.85927   144.87904        Western Metropolitan   
23670     436.0  -37.85274   144.88738        Western Metropolitan   
23672     866.0  -37.85908   144.89299        Western Metropolitan   
23676     362.0  -37.81188   144.88449        Western Metropolitan   

       Propertycount  
1             4019.0  
2             4019.0  
4             4019.0  
5             4019.0  
6             4019.0  
...              ...  
23668         7392.0  
23669         6380.0  
23670         6380.0  
23672         6380.0  
23676         6543.0  

[13603 rows x 18 columns]

##### Now we have removed that rows , which are having missing Landsize values

In [38]:
df1.isnull().sum().sort_values(ascending=False)/len(df1)*100

Car              0.463133
Longtitude       0.161729
Lattitude        0.161729
Bathroom         0.007351
Propertycount    0.000000
SellerG          0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Method           0.000000
Distance         0.000000
Date             0.000000
Regionname       0.000000
Postcode         0.000000
Bedroom2         0.000000
Landsize         0.000000
Suburb           0.000000
dtype: float64

In [39]:
df1.loc[:,["Longtitude","Lattitude"]].describe()

Longtitude     Lattitude
count  13581.000000  13581.000000
mean     144.995221    -37.809204
std        0.103913      0.079257
min      144.431810    -38.182550
25%      144.929600    -37.856820
50%      145.000100    -37.802360
75%      145.058320    -37.756400
max      145.526350    -37.408530

In [40]:
print("Longtitude Mean value",df1["Longtitude"].mean())
print("Lattitude Mean value",df1["Lattitude"].mean())

Longtitude Mean value 144.995220860761
Lattitude Mean value -37.80920350636918


In [41]:
df1.loc[:,"Lattitude"].fillna(df1["Lattitude"].mean(),inplace=True)
df1.loc[:,"Longtitude"].fillna(df1["Longtitude"].mean(),inplace=True)

#####  We have treated Lattitude & Longtitude missing values with mean of Lattitude & Longtitude attributes

In [42]:
df1.isnull().sum().sort_values(ascending=False)/len(df1)*100

Car              0.463133
Bathroom         0.007351
Propertycount    0.000000
Date             0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Method           0.000000
SellerG          0.000000
Distance         0.000000
Regionname       0.000000
Postcode         0.000000
Bedroom2         0.000000
Landsize         0.000000
Lattitude        0.000000
Longtitude       0.000000
Suburb           0.000000
dtype: float64

In [43]:
df1.loc[:,["Car","Bathroom"]].describe()

Car      Bathroom
count  13540.000000  13602.000000
mean       1.610414      1.534921
std        0.962244      0.691834
min        0.000000      0.000000
25%        1.000000      1.000000
50%        2.000000      1.000000
75%        2.000000      2.000000
max       10.000000      8.000000

In [44]:
df1.groupby("Car").Car.count().sort_values(ascending=False)

Car
2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [45]:
df1.groupby("Bathroom").Car.count().sort_values(ascending=False)

Bathroom
1.0    7472
2.0    4971
3.0     920
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Car, dtype: int64

In [46]:
df1.loc[:,"Car"].fillna(2,inplace=True)
df1.loc[:,"Bathroom"].fillna(1,inplace=True)

##### We have treated Car & Bathroom missing values with 2 &1 in  Car & Bathroom attributes

In [47]:
df1.isnull().sum().sort_values(ascending=False)/len(df1)*100

Propertycount    0.0
Regionname       0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Suburb           0.0
dtype: float64

### <b><font color=dodgerblue> Q14. Now since you have completed your data cleaning, determine the number and percentage of columns removed.
### Ans :-

In [48]:
df1.shape

(13603, 18)

In [49]:
print("Percentage of colummns we have removed in data cleaning :-",len(col)/len(Original_data.columns) *100)

Percentage of colummns we have removed in data cleaning :- 14.285714285714285


### <b><font color=dodgerblue> Q15. Determine the number and percentage of rows removed in whole process. Also comment briefly that according to you the amount of data lost is justified or not.
### Ans :-

In [50]:
rows_removed = len(Original_data) - len(df1)
rows_removed

10075

In [51]:
print("Total rows removed in data cleaning:-",rows_removed/len(Original_data) *100)

Total rows removed in data cleaning:- 42.550046456626404


# ========
# Method 2
# ========

### <b><font color=dodgerblue> Q12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? Give a brief explanation and treat the missing values in 'Price' column.
### Ans :-

In [52]:
df2 = copy.deepcopy(df)
df.shape

(19269, 18)

In [53]:
new_df = df2.interpolate()
new_df

Suburb             Address  Rooms Type         Price Method  \
0        Abbotsford       68 Studley St    2.0    h           NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1.480000e+06      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1.035000e+06      S   
3        Abbotsford  18/659 Victoria St    3.0    u  1.250000e+06     VB   
4        Abbotsford        5 Charles St    3.0    h  1.465000e+06     SP   
...             ...                 ...    ...  ...           ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u  2.033333e+06     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h  1.566667e+06      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1.100000e+06      S   
23676    Yarraville          6 Agnes St    4.0    h  1.285000e+06     SP   
23677    Yarraville       33 Freeman St    4.0    h  1.050000e+06     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0       3.5       2.0  1.0   
23675     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

         Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0      126.000000  -37.80140   144.99580  Northern Metropolitan         4019.0  
1      202.000000  -37.79960   144.99840  Northern Metropolitan         4019.0  
2      156.000000  -37.80790   144.99340  Northern Metropolitan         4019.0  
3        0.000000  -37.81140   145.01160  Northern Metropolitan         4019.0  
4      134.000000  -37.80930   144.99440  Northern Metropolitan         4019.0  
...           ...        ...         ...                    ...            ...  
23673  552.000000  -37.90032   144.61839   Western Metropolitan         5262.0  
23674  488.666667  -37.87882   144.60184   Western Metropolitan         5262.0  
23675  425.333333  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676  362.000000  -37.81188   144.88449   Western Metropolitan         6543.0  
23677  362.000000  -37.81829   144.87404   Western Metropolitan         6543.0  

[19269 rows x 18 columns]

### Treating all missing values with the interpolate function

In [54]:
new_df.shape

(19269, 18)

In [55]:
new_df.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            1
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

##### Now we can see, there is only one null values is present in dataset. So we will remove it.

In [56]:
new_df = new_df[new_df["Price"].notnull()]
new_df

Suburb             Address  Rooms Type         Price Method  \
1        Abbotsford        85 Turner St    2.0    h  1.480000e+06      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1.035000e+06      S   
3        Abbotsford  18/659 Victoria St    3.0    u  1.250000e+06     VB   
4        Abbotsford        5 Charles St    3.0    h  1.465000e+06     SP   
5        Abbotsford    40 Federation La    3.0    h  8.500000e+05     PI   
...             ...                 ...    ...  ...           ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u  2.033333e+06     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h  1.566667e+06      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1.100000e+06      S   
23676    Yarraville          6 Agnes St    4.0    h  1.285000e+06     SP   
23677    Yarraville       33 Freeman St    4.0    h  1.050000e+06     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
...              ...         ...       ...       ...       ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0       3.5       2.0  1.0   
23675     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

         Landsize  Lattitude  Longtitude             Regionname  Propertycount  
1      202.000000  -37.79960   144.99840  Northern Metropolitan         4019.0  
2      156.000000  -37.80790   144.99340  Northern Metropolitan         4019.0  
3        0.000000  -37.81140   145.01160  Northern Metropolitan         4019.0  
4      134.000000  -37.80930   144.99440  Northern Metropolitan         4019.0  
5       94.000000  -37.79690   144.99690  Northern Metropolitan         4019.0  
...           ...        ...         ...                    ...            ...  
23673  552.000000  -37.90032   144.61839   Western Metropolitan         5262.0  
23674  488.666667  -37.87882   144.60184   Western Metropolitan         5262.0  
23675  425.333333  -37.72006   145.10547  Northern Metropolitan         1369.0  
23676  362.000000  -37.81188   144.88449   Western Metropolitan         6543.0  
23677  362.000000  -37.81829   144.87404   Western Metropolitan         6543.0  

[19268 rows x 18 columns]

In [57]:
new_df.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

### <b><font color=dodgerblue> Q13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most suitable method.
### Ans :-

In [58]:
new_df.isnull().sum().sort_values(ascending=False)/len(new_df)*100

Propertycount    0.0
Regionname       0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Suburb           0.0
dtype: float64

##### There is zero percentage of null value is present in data, so we will further move for visualization

### <b><font color=dodgerblue> Q14. Now since you have completed your data cleaning, determine the number and percentage of columns removed.
### Ans :-

In [59]:
new_df.shape

(19268, 18)

In [60]:
print("Percentage of colummns we have removed in data cleaning :-",len(col)/len(Original_data.columns) *100)

Percentage of colummns we have removed in data cleaning :- 14.285714285714285


### <b><font color=dodgerblue> Q15. Determine the number and percentage of rows removed in whole process. Also comment briefly that according to you the amount of data lost is justified or not.
### Ans :-

In [61]:
rows_r = len(Original_data) - len(new_df)
rows_r

4410

In [62]:
print("Total rows removed in data cleaning:-",rows_r/len(Original_data) *100)

Total rows removed in data cleaning:- 18.62488385843399


## <b><font color=red> So, from Method 1 we removed 14.2857 % columns & 42.5500 % rows & from Method 2 we lost 14.2857 % columns & 18.6248 % rows. As we can see in method 1 there is big loss in data almost 43%, so that is not suitable or to predict any thing from remaining data. But there is a hope from method 2, we lost here almost 19% data, but if we use some other packages or someother functions. So we can control our data loss under 10%. Actually data is really precious, if we want prediction more accurately or precision from models, so we can't loose more than 7 to 8 % data.
    